lib/handler/pool/state.ex

```
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
-    if has_worker_named(state, task_name) do
-      {:ok, number_killed} = Pool.kill(pool, task_name)
-      {:ok, number_killed, state}
-    else
-      {:ok, 0, state}
-    end
+  @spec kill_worker(t(), String.t()) :: {:ok, t(), non_neg_integer()}
+  def kill_worker(state, task_name) do
+    Enum.reduce(state.workers, {:ok, state, 0}, fn
+      {_ref, %{task_pid: task_pid, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        :ok = Process.send(task_pid, {Pool, :user_killed}, [])
+        {:ok, state, number_killed + 1}
+
+      {ref, %{delegated_to: pool, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        case Pool.kill_by_ref(pool, ref) do
+          :ok ->
+            {:ok, state, number_killed + 1}
+
+          :no_such_worker ->
+            {:ok, state, number_killed}
+        end
+
+      {_ref, _worker}, {:ok, state, number_killed} ->
+        {:ok, state, number_killed}
+    end)
   end
 
-  def kill_worker(state, task_name) do
-    {state, number_killed} =
-      Enum.reduce(state.workers, {state, 0}, fn
-        {ref, {_bytes_commited, _from_pid, task_pid, ^task_name}}, {state, number_killed} ->
-          Process.unlink(task_pid)
-          Process.exit(task_pid, :user_killed)
-
-          exception =
-            Handler.ProcessExit.exception(
-              message: "User killed the process",
-              reason: :user_killed
-            )
-
-          state =
-            state
-            |> send_response(ref, {:error, exception})
-            |> cleanup_commitments(ref)
-
-          {state, number_killed + 1}
-
-        _other_worker, {state, number_killed} ->
-          {state, number_killed}
-      end)
+  def kill_worker_by_ref(%State{workers: workers}, ref) do
+    case Map.get(workers, ref) do
+      %{delegated_to: pool} ->
+        Pool.kill_by_ref(pool, ref)
 
-    {:ok, number_killed, state}
+      %{task_pid: pid} ->
+        :ok = Process.send(pid, {Pool, :user_killed}, [])
+        :ok
+    end
   end
 
```

Under what circumstances would `Process.send/3` fail, and why would blowing up be the right thing to do?

---

That's a great question. When as I was writing this the only use-case I could imagine where it would fail would be a race condition where a process completes and exits as we are in the process of trying to kill it. I think that race condition will be pretty rare since these `kill` operators are pretty fast so it would be a narrow window of time.

But, message delivery is not explicitly guaranteed by the VM and I'm sure it can fail for other reasons such as network interruptions. I wasn't sure of the best way to handle those failures, so I decided that having the process crash was the the safest option given that this was a pretty unknown space of potential problems.

TL/DR; If I know which errors to account for I'll try to recover, but when I don't know how to recover I follow the "let it crash" philosophy because that will force the process to "turn it off and turn it back on" to get a nice clean state